includes/class-amp-wp-styles.php

```diff
@@ -0,0 +1,216 @@
+<?php
+/**
+ * Class AMP_WP_Styles
+ *
+ * @package AMP
+ */
+
+/**
+ * Extend WP_Styles with handling of CSS for AMP.
+ *
+ * @since 0.7
+ */
+class AMP_WP_Styles extends WP_Styles {
+
+	/**
+	 * Concatenation is always enabled for AMP.
+	 *
+	 * @since 0.7
+	 * @var bool
+	 */
+	public $do_concat = true;
+
+	/**
+	 * Generates an enqueued style's fully-qualified file path.
+	 *
+	 * @since 0.7
+	 * @see WP_Styles::_css_href()
+	 *
+	 * @param string $src The source URL of the enqueued style.
+	 * @param string $handle The style's registered handle.
+	 * @return string|WP_Error Style's absolute validated filesystem path, or WP_Error when error.
+	 */
+	public function get_validated_css_file_path( $src, $handle ) {
+		if ( ! is_bool( $src ) && ! preg_match( '|^(https?:)?//|', $src ) && ! ( $this->content_url && 0 === strpos( $src, $this->content_url ) ) ) {
+			$src = $this->base_url . $src;
+		}
+
+		/** This filter is documented in wp-includes/class.wp-styles.php */
+		$src = apply_filters( 'style_loader_src', $src, $handle );
+
+		// Strip query and fragment from URL.
+		$src = preg_replace( ':[\?#].+:', '', $src );
+
+		$src = esc_url_raw( $src );
+
+		// @todo Explicitly not using includes_url() or content_url() since filters may point outside filesystem?
+		$includes_url = includes_url( '/' );
+		$content_url  = content_url( '/' );
+		$admin_url    = get_admin_url( null, '/' );
+		$css_path     = null;
+		if ( 0 === strpos( $src, $content_url ) ) {
```

Comment:
What if the assets are in root folder? For instance could use the filesystem to create dir and cache all CSS which is then enqueued. [Here here is a function](https://github.com/Getbeans/Beans/blob/development/lib/api/utilities/functions.php#L174) which converts url to path, has be perfected over the years to fix all bugs reported on all sorts of various WP sites structure.

Reply:
No assets from core, themes or plugins would be in the root folder. So I don't think we should accommodate them. Since we're actually getting the file source from the filesystem (rather than relying on whether the requesting user has HTTP access) and outputting it, I think we should be very restrictive about what we allow. For example, if there was a `/private/secrets.js` or something, we would not want this file to be enqueued. True we're talking about CSS here, but still.

Reply:
Sounds good, I attempted to test on Windows as I recalled issues when doing URL to path replacements but gave up due to issues encountered when trying to run a stable local install on my virtual machine. I will leave that task to QA (Claudio).